### PR TITLE
KaizenTask768_Add_smoke_test_for_lib_tasks_run_model_experiment_notebooks

### PR DESCRIPTION
**Pull request by Jd8111997**
Hi @sonaalKant,

I've written a unit test for the `run_notebooks` method. Could you please review the logic of the function and let me know if I'm heading in the right direction?

Also, I want to mention that this is just a draft PR, so the code hasn't been refactored yet.

While running the pytest, I encountered the following error:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.9/unittest/case.py", line 593, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.9/unittest/case.py", line 550, in _callTestMethod
    method()
  File "/app/dev_scripts/test/test_run_notebook.py", line 574, in test1
    run_experiment_notebooks.run_notebooks(
  File "/venv/lib/python3.9/site-packages/invoke/tasks.py", line 134, in __call__
    if not isinstance(args[0], Context):
IndexError: tuple index out of range
```

I am mocking a Context object and passing it to the `run_notebooks` method. Do you have any insights into why it's failing? Since I couldn't run this on real-world data, could you please help me with this?

### CODE REVIEW

**Review by gpsaggese — COMMENTED** [review]
Is it really needed to mock so much? We want to test end-to-end. 
Can we just create a notebook and run the actual code? 

@sonaalKant do you see what I mean?

**Comment by sonaalKant** [comment]
This was not the expectation at all, mocking everything destroys the purpose of having a unit test. We want to test the flow, if there are any changes in config or in the function reading the config then the test should be able to catch this. 

You only need two unit tests one for broker_only and one for full system: 

Steps:
 1. Create a temp directory and store the config file (created by you) in that temp directory. (args.json for broker only run and pkl file for full system ) 
 2. Mock _run_notebook function to return nothing

Also, `test_run_notebooks.py` is not the correct file to place your code. Go through the documentation again https://github.com/cryptokaizen/cmamp/blob/e57bd0e992a82e40f8e48fa977c105bd2b4f45b6/docs/coding/all.write_unit_tests.how_to_guide.md#naming-and-placement-conventions if things are not clear.

**Comment by Jd8111997** [comment]
@sonaalKant, I updated a PR, could you please review it and let me know if I am heading into the right direction.

Also, What's the use of `ctx` object in `run_notebooks` method ?
I see we are passing it to the method but not using it. I tried to pass `None` but I am getting the error and I also tried to mock it but still the same error persists. Let me know if you have any solution in mind.

This is the error I'm getting:
```
_____________________________________________ Test_run_notebooks.test1 _____________________________________________
Traceback (most recent call last):
  File "/usr/lib/python3.9/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.9/unittest/case.py", line 593, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.9/unittest/case.py", line 550, in _callTestMethod
    method()
  File "/app/dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py", line 48, in test1
    run_experiment_notebooks.run_notebooks(
  File "/venv/lib/python3.9/site-packages/invoke/tasks.py", line 134, in __call__
    if not isinstance(args[0], Context):
IndexError: tuple index out of range
```

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    def tear_down_test(self) -> None:
+        self.mock__run_notebook.stop()
+        self.mock_publish_notebook.stop()
+
+    def test1(self) -> None:
+        """
+        Broker only run.
+        """
+        with tempfile.TemporaryDirectory() as temp_dir:
```
Comment (sonaalKant) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:38]: You dont have to use tmpfile, we already have inbuilt capability of using scratch directory. https://github.com/cryptokaizen/cmamp/blob/fceef579df860e5c58007c66f924e0bba13be720/helpers/hunit_test.py#L1196

Comment (sonaalKant) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:38]: You dont have to use tmpfile, we already have inbuilt capability of using scratch directory. https://github.com/cryptokaizen/cmamp/blob/fceef579df860e5c58007c66f924e0bba13be720/helpers/hunit_test.py#L1196

File: dev_scripts/test/test_run_notebook.py [hunk 2]
```diff
@@ -12,7 +12,9 @@
 import helpers.hprint as hprint
 import helpers.hsystem as hsystem
 import helpers.hunit_test as hunitest
-
+import unittest.mock as umock
```
Comment (sonaalKant) [inline dev_scripts/test/test_run_notebook.py:15]: remove all the changes from this file.

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+            with open(os.path.join(temp_dir, "args.json"), "w") as f:
+                json.dump(broker_only_config, f)
+
+            run_experiment_notebooks.run_notebooks(
+                ctx = self.ctx,
+                system_log_dir=temp_dir,
+                base_dst_dir="/notebooks_output"
+            )
+            self.run__notebook_mock.assert_called_once()
```
Comment (sonaalKant) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:53]: this is not required.

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            broker_only_config = {
+                "parent_order_duration_in_min": 32,
+                "universe" : "test_universe",
+                "child_order_execution_freq": "hourly",
+            }
+
+            with open(os.path.join(temp_dir, "args.json"), "w") as f:
+                json.dump(broker_only_config, f)
+
+            run_experiment_notebooks.run_notebooks(
```
Comment (sonaalKant) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:48]: You have to do a cmd line call to test this method. refer: 
https://github.com/cryptokaizen/cmamp/blob/fceef579df860e5c58007c66f924e0bba13be720/dev_scripts/test/test_run_notebook.py

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+from typing import Any, Dict, Iterable, List, Tuple, Generator
+import unittest.mock as umock
+from invoke import Context
+import pytest
+import tempfile
+import os
+import json
+import pickle
+import helpers.hunit_test as hunitest
+import dev_scripts.lib_tasks_run_model_experiment_notebooks as run_experiment_notebooks
```
Comment (sonaalKant) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:10]: Thats not how we import stuff.

**Review by sonaalKant — CHANGES_REQUESTED** [review]

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py (hunk 5 above)
Comment (Jd8111997) [reply dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:10]: Yes, I know, this is just a draft PR. I am just figuring out to write correct test cases.

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py (hunk 4 above)
Comment (Jd8111997) [reply dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:48]: Aren't we want to test only a single `run_notebooks` method. As I see, this method is used to run some python script using command line. What's the purpose of using this ?

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+
```
Comment (gpsaggese) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:1]: Run linter to clean up the code

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import pickle
+import pytest
+from invoke import Context
+import unittest.mock as umock
+import helpers.hunit_test as hunitest
+from typing import Any, Dict, Iterable, List, Tuple, Generator
+import dev_scripts.lib_tasks_run_model_experiment_notebooks as dsltrmen
+
+class Test_run_notebooks(hunitest.TestCase):
+    mock__run_notebook = umock.patch.object(dsltrmen, "_run_notebook")
```
Comment (gpsaggese) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:13]: I don't understand this. The goal is to call this actual function so we can test that works.

There should be no mocking in this unit test.

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        scratch_space = self.get_scratch_space(test_class_name=self.__class__.__name__,
+                                               test_method_name=self._testMethodName)
+        broker_only_config = {
+            "parent_order_duration_in_min": 32,
+            "universe" : "test_universe",
+            "child_order_execution_freq": "hourly",
+        }
+
+        with open(os.path.join(scratch_space, "args.json"), "w") as f:
+            json.dump(broker_only_config, f)
```
Comment (gpsaggese) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:47]: I think we have some these functions in helpers/hpickle.py

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            ctx = self.ctx,
+            system_log_dir=scratch_space,
+            base_dst_dir="/notebooks_output"
+        )
+
+    def test2(self) -> None:
+        """
+        full system config run.
+        """
+        scratch_space = self.get_scratch_space(test_class_name=self.__class__.__name__,
```
Comment (gpsaggese) [inline dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:60]: A PR with 0 comments?

Comment (Jd8111997) [reply dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:60]: @gpsaggese This is just a draft PR at the moment. Currently, I'm concentrating on devising accurate unit test cases so I can push more changes and receive prompt feedback.

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py (hunk 8 above)
Comment (Jd8111997) [reply dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:47]: @gpsaggese Thanks for suggestions, I will take a look.

File: dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py (hunk 7 above)
Comment (Jd8111997) [reply dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:13]: @gpsaggese okay, I have been told to mock `_run_notebook` and `publish_system_reconciliation_notebook` function. You can check the previous comment. I will discuss this with @sonaalKant.

Comment (sonaalKant) [reply dev_scripts/test/test_lib_tasks_run_model_experiment_notebooks.py:13]: @gpsaggese This is a simple unit test just to verify the flow. Context is, changes made on system config side  (reading and manipulating configs) results in occasional failure of this method as there are no unit test. We don't want to test `run_notebook` as it is already been tested we want only want to test the flow.

**Comment by Jd8111997** [comment]
> One suggestion from my side would be to shift this PR on cmamp side so Jaydeep can more clarity and it can be tested properly.
> 
> Would be nice if Sonaal can point Jaydeep to some similar test case for reference.
> 
> If we don't have any, a PP with Jaydeep can make things more clear I believe.

@sonaalKant, this might help.

**Comment by sonaalKant** [comment]
> One suggestion from my side would be to shift this PR on cmamp side so Jaydeep can more clarity and it can be tested properly.

Agree. @Jd8111997 Once you are done can you close this PR here and create a final PR on cmamp directly. https://github.com/cryptokaizen/cmamp/issues/7782

**Comment by Jd8111997** [comment]
@sonaalKant, there is a bug in the `run_notebooks` function. We are logging the `price_col` and `table_name` variables outside of the `if.. else..` statement, but they are only defined for a `full_system_run`. Therefore, in the `broker_only` scenario, it throws the following error:

```
UnboundLocalError: local variable 'price_col' referenced before assignment
```

**Comment by sonaalKant** [comment]
Good catch. Thats exactly why we need unit test to check the flow. I see some changes were made 3 weeks back but this case got missed.

**Comment by sonaalKant** [comment]
Shifted to cmamp and merged https://github.com/cryptokaizen/cmamp/issues/7782. Closing here

**Pull request closed without merging**